### PR TITLE
⭐️ cpe for assets and packages

### PR DESCRIPTION
**Pull request by chris-rock**
This PR adds CPE support for operating systems and packages. `asset` and to `packages` as well as `python.package` have a new `cpes` field. Since CPEs are not consistently used, we designed it to allow us to return multiple cpes. There is a question about versions, architecture etc. We anticipate to add more CPE entries to cover the different cases going forward.

**asset**

```coffeescript
cnquery> asset.cpes
asset.cpes: [
  0: cpe uri="cpe:2.3:o:rocky:rocky:9:*:baseos:*:*:*:*:*"
]
```

CPEs are a new resource type and you can inspect the CPE via:

```
cnquery> asset.cpes { * }
asset.cpes: [
  0: {
    targetHw: ""
    product: "rocky"
    swEdition: ""
    uri: "cpe:2.3:o:rocky:rocky:9:*:baseos:*:*:*:*:*"
    edition: "baseos"
    targetSw: ""
    language: ""
    other: ""
    update: ""
    part: "o"
    vendor: "rocky"
    version: "9"
  }
]
```

**os packages**

```coffeescript
cnquery> packages { cpes }
packages.list: [
  0: {
    cpes: [
      0: cpe uri="cpe:2.3:a:libgcc:libgcc:11.3.1-4.3.el9:x86_64:*:*:*:*:*:*"
    ]
  }
  1: {
    cpes: [
      0: cpe uri="cpe:2.3:a:crypto-policies:crypto-policies:20221215-1.git9a18988.el9:noarch:*:*:*:*:*:*"
    ]
  }
  2: {
    cpes: [
      0: cpe uri="cpe:2.3:a:tzdata:tzdata:2023c-1.el9:noarch:*:*:*:*:*:*"
    ]
  }
  ...
]
```

or if you want a flat list of all list:

> Note there are two `.map` since we have potentially multiple cpes for each asset. `cpes.map(uri)` returns plain list of the cpe uri. Then we map all packages to the list of all cpes. Since each package has can have multiple cpes we need `.flat` to make it a plain list

```coffeescript
cnspec> packages.map(cpes.map(uri)).flat
packages.map.flat: [
  0: "cpe:2.3:a:libgcc:libgcc:11.3.1-4.3.el9:x86_64:*:*:*:*:*:*"
  1: "cpe:2.3:a:crypto-policies:crypto-policies:20221215-1.git9a18988.el9:noarch:*:*:*:*:*:*"
  2: "cpe:2.3:a:tzdata:tzdata:2023c-1.el9:noarch:*:*:*:*:*:*"
  3: "cpe:2.3:a:rocky-gpg-keys:rocky-gpg-keys:9.2-1.4.el9:noarch:*:*:*:*:*:*"
  4: "cpe:2.3:a:rocky-release:rocky-release:9.2-1.4.el9:noarch:*:*:*:*:*:*"
  5: "cpe:2.3:a:rocky-repos:rocky-repos:9.2-1.4.el9:noarch:*:*:*:*:*:*"
  6: "cpe:2.3:a:setup:setup:2.13.7-9.el9:noarch:*:*:*:*:*:*"
  7: "cpe:2.3:a:filesystem:filesystem:3.16-2.el9:x86_64:*:*:*:*:*:*"
  8: "cpe:2.3:a:basesystem:basesystem:11-13.el9:noarch:*:*:*:*:*:*"
  9: "cpe:2.3:a:libreport-filesystem:libreport-filesystem:2.15.2-6.el9.rocky.0.2:noarch:*:*:*:*:*:*"
```

**python.packages**

```coffeescript
cnquery> python.packages { cpes }
python.packages: [
  0: {
    cpes: [
      0: cpe uri="cpe:2.3:a:gpg_project:gpg:1.15.1:*:*:*:*:*:*:*"
    ]
  }
  1: {
    cpes: [
      0: cpe uri="cpe:2.3:a:libcomps_project:libcomps:0.1.18:*:*:*:*:*:*:*"
    ]
  }
  2: {
    cpes: [
      0: cpe uri="cpe:2.3:a:rpm_project:rpm:4.16.1.3:*:*:*:*:*:*:*"
    ]
  }
]
```

In addition you can also create your own cpe resources and get the data:

```
cnquery> cpe("cpe:2.3:a:microsoft:internet_explorer:8.0.6001:beta:*:*:*:*:*:*") { * }
cpe: {
  swEdition: ""
  vendor: "microsoft"
  product: "internet_explorer"
  other: ""
  part: "a"
  edition: ""
  uri: "cpe:2.3:a:microsoft:internet_explorer:8.0.6001:beta:*:*:*:*:*:*"
  version: "8.0.6001"
  targetSw: ""
  language: ""
  targetHw: ""
  update: "beta"
}
```

Since this resource depends on the core resources which are bundled in cnquery, users need to use latest cnquery to access the resources. Otherwise the following error will be shown:

```
cnquery> asset.cpes
rpc error: code = Unknown desc = cannot find resource 'cpe' in schema
asset.cpes: null
```

### CODE REVIEW

File: .github/actions/spelling/expect.txt [hunk 1]
```diff
@@ -11,6 +11,7 @@ certificatechains
 cmek
 Cooldown
 cryptokey
+cpes
```
Comment (tas50) [inline .github/actions/spelling/expect.txt]: We should stick this in the main dictionary. It's going to show up all over.

Comment (chris-rock) [reply .github/actions/spelling/expect.txt]: agreed, would love to do this as a separate task

Comment (tas50) [reply .github/actions/spelling/expect.txt]: It's in the main dictionary now

Comment (chris-rock) [reply .github/actions/spelling/expect.txt]: Merci, I am going to remove it from here then

**Comment by arlimus** [comment]
A thought on the direction of CPEs: We are currently taking a string-based approach here, which is nice and lightweight, but limited in terms of accessing the data encoded inside the string. If we decide to access it, we will run into type-changes in the future, which are hard to keep compatible.

(There is a potential route to go via type-inheritance of the same base type (eg string), but this is still hypothetical at this point...)

I'd recommend going with a structured type for this field instead and start incorporating accessors to the URI if we can

**Comment by chris-rock** [comment]
Great idea @arlimus, I'll adjust the PR and introduce a new cve type.

**Comment by chris-rock** [comment]
@arlimus I added a new `cpe` type and update the implementation to return the cpe type instead of plain string values.

**Review by arlimus — APPROVED** [review]
This is a super amazing addition! Say what you will about CPEs, but having them in cnquery/cnspec is :rocket: !!

**Pull request merged**